core/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -445,13 +442,22 @@ private void checkDepthLimit(String objectPath, long maxDepth) {
         }
     }
 
-    public DocumentMapper parse(String mappingType, CompressedXContent mappingSource, boolean applyDefault) throws MapperParsingException {
-        String defaultMappingSource;
-        if (PercolatorFieldMapper.TYPE_NAME.equals(mappingType)) {
-            defaultMappingSource = this.defaultPercolatorMappingSource;
-        }  else {
-            defaultMappingSource = this.defaultMappingSource;
+    /**
+     * We only allow upto 1 percolator field per index.
+     *
+     * Reasoning here is that the PercolatorQueryCache only supports a single document having a percolator query.
+     * Also specifying multiple queries per document feels like an anti pattern
+     */
+    private void checkPercolatorFieldLimit(List<FieldMapper> fieldMappers) {
+        long numPercolatorFields = fieldMappers.stream()
+                .filter(fieldMapper -> fieldMapper instanceof PercolatorFieldMapper)
+                .count();
+        if (numPercolatorFields > 1) {
+            throw new IllegalArgumentException("Up to one percolator field type is allowed per index");
         }
+    }
```

Comment:
Can you iterate fieldTypes (which contains all index level field types rather than type level mappers) instead?